src/bgw/scheduler.c

```diff
@@ -247,6 +249,7 @@ scheduled_bgw_job_transition_state_to(ScheduledBgwJob *sjob, JobState new_state)
 				elog(WARNING,
 					 "scheduler detected that job %d was deleted when starting job",
 					 sjob->job.fd.id);
+				ts_bgw_job_cache_invalidate_callback();
 				scheduled_bgw_job_transition_state_to(sjob, JOB_STATE_SCHEDULED);
```

Comment:
I know this is not a change in this PR, but why do we recursively try to transition to `JOB_STATE_SCHEDULED` when we've already asserted above that the current state is already  `JOB_STATE_SCHEDULED`?

This seems to create an infinite loop if the same error happens again (I assume the job won't magically appear in the `bgw_job` table, so the same issue is likely to happen in the recursive call).

Reply:
This will go into scheduled state and set up the next_start time for the job. The bgw scheduler makes 1 pass over all the jobs and then updates its cache if needed (ts_bgw_...cache_invalidate...callback will set it as needed). So in the next pass , it will not find this job if it has indeed been deleted.

Reply:
By the way, the  recursive loop is what we have been running into. so we missed the ts_catalog_cache_invalidate callback somehow.

Reply:
I guess I don't understand why we at all try again if the job was deleted. Don't we trust that this is the case the first time around?

FWIW, `ts_bgw_job_get_share_lock` seems to scan the jobs table so it _should_ be trustworthy w.r.t. whether this job exists or not so it doesn't make sense to me to try to schedule it again just to fail. The fact that the cache was not invalidated is another issue.

Reply:
we are not trying again here. The bgw scheduler runs in a continuous loop (ts_bgw_scheduler_process). This happens during 1 pass of the scheduler loop. For the next pass of the loop,  the scheduler should update the list of cached jobs (and remove this job from its cache) and so we should not run into it the next time around.

Reply:
We are calling `scheduled_bgw_job_transition_state_to(sjob, JOB_STATE_SCHEDULED)` on a job that doesn't exist and is already in  `JOB_STATE_SCHEDULED`. Isn't that trying again? In any case, the action taken here is really strange.

This is what is happening here, if I am not mistaken:

1. This function was called with sjob in state `JOB_STATE_SCHEDULED` (asserted) and trying to transition to `JOB_STATE_STARTED`.
2. The job was not found (or lock could not be taken), so we got here
3. The sjob is still in state `JOB_STATE_SCHEDULED` since we haven't transitioned yet and now we are trying to transition to the same state we are in, i.e., `JOB_STATE_SCHEDULED` to `JOB_STATE_SCHEDULED`.
4. Presumably, the steps above will eventually lead to a transition to `JOB_STATE_STARTED` again, which means we are back to step (1). Or, alternatively, the job doesn't exist and will fail somehow when rescheduled, but we knew that already in step (2).